File: src/compile.js
import { parsePattern } from "./pattern.js";

/**
 * @typedef {Object} CompiledRoutes
 * @property {Map<string, {route: string, value: any}>} staticRoutes - Routes with only literal segments, keyed by pattern.
 * @property {Array<{route: string, value: any, segments: import("./pattern.js").PatternSegment[], paramNames: string[]}>} dynamicRoutes - Routes containing parameters, in definition order.
 * @property {{route: string, value: any} | null} wildcard - The optional "*" fallback route.
 */

/**
 * @typedef {Object} MatchResult
 * @property {boolean} found - Whether a route was matched.
 * @property {string} pathname - The normalized pathname that was matched against.
 * @property {string | null} route - The matched route pattern, or null when not found.
 * @property {Object<string, string>} params - Extracted route parameters.
 * @property {any} value - The opaque value associated with the matched route.
 */

/**
 * @typedef {Object} MatchStrategy
 * @property {(pathname: string) => MatchResult} match - Resolve a pathname to a match result.
 */

/**
 * Compile route definitions into an internal, reusable representation.
 *
 * Routes are partitioned into static routes (exact string lookup), dynamic
 * routes (segments containing parameters), and an optional wildcard fallback.
 * Compilation happens once; the resulting object is reused for every match.
 *
 * @param {Object<string, any>} routes - Map of pattern strings to opaque values.
 * @returns {CompiledRoutes} The compiled route collection.
 */
export function compile(routes) {
  const staticRoutes = new Map();
  const dynamicRoutes = [];
  let wildcard = null;

  for (const [pattern, value] of Object.entries(routes)) {
    const parsed = parsePattern(pattern);

    if (parsed.isWildcard) {
      wildcard = { route: pattern, value };
      continue;
    }

    const onlyStatic = parsed.segments.every((segment) => segment.type === "static");

    if (onlyStatic) {
      staticRoutes.set(parsed.raw, { route: pattern, value });
      continue;
    }

    dynamicRoutes.push({
      route: pattern,
      value,
      segments: parsed.segments,
      paramNames: parsed.segments
        .filter((segment) => segment.type === "param")
        .map((segment) => segment.name),
    });
  }

  return { staticRoutes, dynamicRoutes, wildcard };
}

File: go/compile.go
package roution

// StaticRoute is a compiled static route entry.
type StaticRoute struct {
	Route string
	Value any
}

// DynamicRoute is a compiled dynamic route entry.
type DynamicRoute struct {
	Route      string
	Value      any
	Segments   []Segment
	ParamNames []string
}

// CompiledRoutes is the internal, reusable representation of route definitions.
type CompiledRoutes struct {
	StaticRoutes map[string]StaticRoute
	DynamicRoutes []DynamicRoute
	Wildcard     *StaticRoute
}

// Compile partitions route definitions into static routes, dynamic routes, and
// an optional wildcard fallback. Compilation happens once; the result is reused
// for every match.
func Compile(routes map[string]any) CompiledRoutes {
	staticRoutes := make(map[string]StaticRoute)
	var dynamicRoutes []DynamicRoute
	var wildcard *StaticRoute

	for pattern, value := range routes {
		parsed := ParsePattern(pattern)

		if parsed.IsWildCard {
			w := StaticRoute{Route: pattern, Value: value}
			wildcard = &w
			continue
		}

		onlyStatic := true
		for _, seg := range parsed.Segments {
			if seg.Type != SegmentStatic {
				onlyStatic = false
				break
			}
		}

		if onlyStatic {
			staticRoutes[parsed.Raw] = StaticRoute{Route: pattern, Value: value}
			continue
		}

		var paramNames []string
		for _, seg := range parsed.Segments {
			if seg.Type == SegmentParam {
				paramNames = append(paramNames, seg.Name)
			}
		}

		dynamicRoutes = append(dynamicRoutes, DynamicRoute{
			Route:      pattern,
			Value:      value,
			Segments:   parsed.Segments,
			ParamNames: paramNames,
		})
	}

	// Go maps do not preserve insertion order, so sort dynamic routes by
	// pattern for deterministic matching priority (first match wins).
	sortDynamicRoutes(dynamicRoutes)

	return CompiledRoutes{
		StaticRoutes: staticRoutes,
		DynamicRoutes: dynamicRoutes,
		Wildcard:     wildcard,
	}
}

func sortDynamicRoutes(routes []DynamicRoute) {
	// Simple insertion sort keeps the dependency small (no sort import needed
	// for typical route counts) while guaranteeing a stable, reproducible order.
	for i := 1; i < len(routes); i++ {
		for j := i; j > 0 && routes[j].Route < routes[j-1].Route; j-- {
			routes[j], routes[j-1] = routes[j-1], routes[j]
		}
	}
}
